xdsl/rewriting/composable_rewriting/immutable_ir/immutable_ir.py

```diff
@@ -0,0 +1,607 @@
+from __future__ import annotations
+from abc import ABC
+from dataclasses import dataclass
+from typing import Iterable, Sequence, SupportsIndex, Type, TypeGuard, Any, TypeVar, List
+from xdsl.ir import Attribute, Block, BlockArgument, OpResult, Operation, Region, SSAValue
+
+_T = TypeVar('_T')
+
+
+class IList(List[_T]):
+    """
+    A list that can be frozen. Once frozen, it can not be modified. 
+    In comparison to FrozenList this supports pattern matching.
+    """
+    _frozen: bool = False
+
+    def freeze(self):
+        self._frozen = True
+
+    def _unfreeze(self):
+        self._frozen = False
+
+    def append(self, __object: _T) -> None:
+        if self._frozen:
+            raise Exception("frozen list can not be modified")
+        return super().append(__object)
+
+    def extend(self, __iterable: Iterable[_T]) -> None:
+        if self._frozen:
+            raise Exception("frozen list can not be modified")
+        return super().extend(__iterable)
+
+    def insert(self, __index: SupportsIndex, __object: _T) -> None:
+        if self._frozen:
+            raise Exception("frozen list can not be modified")
+        return super().insert(__index, __object)
+
+    def remove(self, __value: _T) -> None:
+        if self._frozen:
+            raise Exception("frozen list can not be modified")
+        return super().remove(__value)
+
+    def pop(self, __index: SupportsIndex = ...) -> _T:
+        if self._frozen:
+            raise Exception("frozen list can not be modified")
+        return super().pop(__index)
+
+    def clear(self) -> None:
+        if self._frozen:
+            raise Exception("frozen list can not be modified")
+        return super().clear()
+
+    def __setitem__(self, __index: SupportsIndex, __object: _T) -> None:
+        if self._frozen:
+            raise Exception("frozen list can not be modified")
+        return super().__setitem__(__index, __object)
+
+    def __add__(self, __x: Iterable[_T]) -> IList[_T]:
+        if self._frozen:
+            raise Exception("frozen list can not be modified")
+        if isinstance(__x, IList) and __x._frozen:
+            raise Exception("frozen list can not be modified")
+        return IList(super().__add__(__x))  # type: ignore
+
+    def __iadd__(self, __x: Iterable[_T]) -> IList[_T]:
+        if self._frozen:
+            raise Exception("frozen list can not be modified")
+        return super().__iadd__(__x)  # type: ignore
+
+    def __eq__(self, __o: object) -> bool:
+        if isinstance(__o, IList):
+            return super().__eq__(__o)  # type: ignore
+        return False
+
+
+@dataclass(frozen=True)
+class ISSAValue(ABC):
+    """
+    Represents an immutable SSA variable. An immutable SSA variable is either an operation result
```

Comment:
For these comments I mirrored what we have in `ir.py`, where the names of the other classes are not mentioned in the comments.
 
Do you think this makes the structure easier to understand? I would be happy to add this, but don't really think its necessary with the two classes inheriting from `ISSAValue` directly beneath.

Do you want me to do this for the descriptions of all classes in this file? 
Also if we want this here we should probably add it to `ir.py` as well, no?

Reply:
I think this could help newcomers, and we should probably also use this in our files.
We could add this in a separate PR, to clean them both at the same time, what do you think?

Reply:
Happy for that, but please open an issue so that we do not forget. Sometimes we tend to forget doing these additional stuff

Reply:
This is now tracked here: https://github.com/xdslproject/xdsl/issues/560